asf_heat_pump_suitability/pipeline/prepare_features/property_density.py

```diff
@@ -2,67 +2,76 @@
 from asf_heat_pump_suitability.getters import get_datasets
 
 
-def extend_df_with_property_density(enhanced_epc_df: pl.DataFrame) -> pl.DataFrame:
+def generate_df_property_density() -> pl.DataFrame:
     """
-    Add property density feature to the EPC dataset
-    Args:
-        enhanced_epc_df (pl.DataFrame): EPC dataset with additional features
+    Generate dataframe with property density (households per km2) per LSOA/2011 Data Zone in England and Wales, and Scotland,
+    respectively, using Standard Area Measurement of ‘Land Area’ (Area to Mean High Water Excluding Area of Inland Water).
+
     Returns:
-        pl.DataFrame: EPC dataset with property density feature
+        pl.DataFrame: households per km2 per LSOA/Data Zone in England, Wales, and Scotland
     """
-    # Fill None values with a default value (e.g., 0)
-    enhanced_epc_df = enhanced_epc_df.with_columns(
-        [
-            pl.col("Number of households 2021")
-            .fill_null(0)
-            .alias("Number of households 2021"),
-            pl.col("Land Count (Area in KM2)")
-            .fill_null(0)
-            .alias("Land Count (Area in KM2)"),  # Fill with 1 to avoid division by zero
-        ]
-    )
-    # Calculate and add the new column with the property density, handling division by zero
-    enhanced_epc_df = enhanced_epc_df.with_columns(
-        (
-            pl.when(pl.col("Land Count (Area in KM2)") != 0)
-            .then(
-                pl.col("Number of households 2021") / pl.col("Land Count (Area in KM2)")
-            )
-            .otherwise(None)
-        ).alias("households_per_km2")
-    )
-    # Changes 0 values back to None
-    enhanced_epc_df = replace_zeros_with_none_df(
-        enhanced_epc_df, "Number of households 2021"
-    )
-    enhanced_epc_df = replace_zeros_with_none_df(
-        enhanced_epc_df, "Land Count (Area in KM2)"
-    )
+    ew_df = generate_df_property_density_ew()
+    s_df = generate_df_property_density_s().rename({"DataZone": "lsoa"})
 
-    return enhanced_epc_df
+    return pl.concat([ew_df, s_df])
 
 
-def replace_zeros_with_none_df(df: pl.DataFrame, column_name: str) -> pl.DataFrame:
+def generate_df_property_density_ew() -> pl.DataFrame:
     """
-    This function replaces 0 values in the specified column of a DataFrame with None.
-
-    Parameters:
-    df (pl.DataFrame): The DataFrame to modify.
-    column_name (str): The name of the column in which to replace 0 values.
+    Generate dataframe with property density (households per km2) per LSOA in England and Wales using Standard Area Measurement of
+    ‘Land Area’ (Area to Mean High Water Excluding Area of Inland Water).
 
     Returns:
-    pl.DataFrame: The modified DataFrame.
+        pl.DataFrame: households per km2 per LSOA in England and Wales
     """
+    households_df = load_transform_df_n_households_ew()
+    area_df = load_transform_df_land_area_ew()
+    df = households_df.join(area_df, how="inner", on="lsoa")
```

Comment:
@lizgzil here is the inner join between households and area which means we don't end up with any rows with 0 value for area. Therefore, we don't need the code to prevent zero divisions.